drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +45,494 @@ extern "C" {
+        .map_err(|_| RawCabooseError::ReadFailed)?;
+
+    // Security considerations:
+    // If there is no caboose, then we may be indexing 0xff padding, or
+    // code/data and padding, or just code/data, and interpreting that as
+    // the caboose size. After an alignment and size check, some values
+    // could still get through. The pathological case would be code/data
+    // that indexes the CABOOSE_MAGIC constant in the code here that is
+    // testing for the caboose. The parts of the image that could then
+    // be accessed are already openly available. There doesn't seem
+    // to be an opportunity for any denial of service.
+    let caboose_magic_offset = image_end_offset.saturating_sub(caboose_size);
+    if ((caboose_magic_offset % U32_SIZE) != 0)
+        || !(IMAGE_HEADER_OFFSET..caboose_size_offset)
+            .contains(&caboose_magic_offset)
+    {
+        return Err(RawCabooseError::MissingCaboose);
+    }
+
+    let caboose_magic = image
+        .read_word(caboose_magic_offset as u32)
+        .map_err(|_| RawCabooseError::MissingCaboose)?;
+
+    if caboose_magic == CABOOSE_MAGIC {
+        let caboose_range = ((caboose_magic_offset as u32) + U32_SIZE)
+            ..(caboose_size_offset as u32);
+        Ok(caboose_range)
+    } else {
+        Err(RawCabooseError::MissingCaboose)
+    }
 }
 
-/// Return the flash storage address range and flash execution address range.
-/// These are only different for the staged stage0 image.
-pub fn image_range(
-    component: RotComponent,
-    slot: SlotId,
-) -> (Range<u32>, Range<u32>) {
-    unsafe {
-        match (component, slot) {
-            (RotComponent::Hubris, SlotId::A) => (
-                __IMAGE_A_BASE.as_ptr() as u32..__IMAGE_A_END.as_ptr() as u32,
-                __IMAGE_A_BASE.as_ptr() as u32..__IMAGE_A_END.as_ptr() as u32,
-            ),
-            (RotComponent::Hubris, SlotId::B) => (
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::A) => (
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::B) => (
-                __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0NEXT_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
+/// Accessor keeps the implementation details of ImageAccess private
+enum Accessor<'a> {
+    // Flash driver, flash device range
+    Flash {
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        span: FlashRange,
+    },
+    Ram {
+        buffer: &'a [u8],
+        span: FlashRange,
```

Comment:
So, a number of places in the access code below appear to rely on buffer.len() being equal to the length of this span. How is this invariant maintained? Would it make more sense to have span be a base address instead, so that the len isn't duplicated (and thus doesn't have to be kept in sync)?